### PR TITLE
Enhancement/ciac 9358/ibm resilient

### PR DESCRIPTION
**Pull request by barryyosi-panw**
## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [CIAC-9358](https://jira-dc.paloaltonetworks.com/browse/CIAC-9358)

## Description
Enhancement for IBM Resillient Systems integration.
Renamed to: **IBM Security QRadar SOAR**

## Must have
- [X] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/CommonScripts/ReleaseNotes/1_15_62.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+
+#### Incident Fields
+
+##### Breach Confirmation
+
+- %%UPDATE_RN%%
```
Comment (thefrieddan1) [inline Packs/CommonScripts/ReleaseNotes/1_15_62.md:6]: Missing RN

File: Packs/IBMResilientSystems/TestPlaybooks/playbook-IBM_QRadar_Soar_Test_Test.yml
Comment (thefrieddan1) [inline Packs/IBMResilientSystems/TestPlaybooks/playbook-IBM_QRadar_Soar_Test_Test.yml:1]: Did you add the TPB to the conf.json file ?

File: Packs/IBMResilientSystems/Scripts/IbmUpdateNote/IbmUpdateNote.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import demistomock as demisto  # noqa: F401
+from CommonServerPython import *  # noqa: F401
+
+
+def update_note(args: Dict[str, Any]) -> CommandResults:
+    note_id = args.get('note_id', '')
+    note_body = args.get('note_body', '')
+    tags = argToList(args.get('tags', ''))
+    remote_incident_id = demisto.incident()['dbotMirrorId']
+    demisto.debug(f'update_note {note_body=} | {remote_incident_id}')
```
Comment (thefrieddan1) [inline Packs/IBMResilientSystems/Scripts/IbmUpdateNote/IbmUpdateNote.py:10]: is remote_incident_id missing a = as well? like note_body?

File: Packs/IBMResilientSystems/Scripts/IbmConvertTasksToTable/IbmConvertTasksToTable_test.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# import IbmConvertCommentsToTable
```
Comment (thefrieddan1) [inline Packs/IBMResilientSystems/Scripts/IbmConvertTasksToTable/IbmConvertTasksToTable_test.py:1]: ?

File: Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py [hunk 4]
```diff
@@ -1,12 +1,13 @@
...
 import demistomock as demisto  # noqa: F401
 from CommonServerPython import *  # noqa: F401
 import json
 import logging
 import time
-
 import urllib3
-
 import resilient
+from resilient.co3 import SimpleClient, SimpleHTTPException
```
Comment (thefrieddan1) [inline Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py]: SimpleHTTPException not accessed.

File: Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py [hunk 5]
```diff
@@ -1,12 +1,13 @@
+import copy
```
Comment (thefrieddan1) [inline Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py]: copy not accessed.

File: Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py [hunk 6]
```diff
@@ -115,92 +126,204 @@
...
+        "description": "The display name of the incident creator.",
+    }
+}
+
+""" CONSTANTS """
+FILE_DOWNLOAD_ERROR_MESSAGE = "<html><head><title>Download error</title></head><body>Download error</body></html>"
+SCRIPT_ENTITIES = "entities"
+DEFAULT_RETURN_LEVEL = "full"
+DEFAULT_RETRIES = 1
+STATUS_NOT_FOUND = 404
```
Comment (thefrieddan1) [inline Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py:179]: Why not use the HTTPStatus ?
from http import HTTPStatus

File: Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py [hunk 7]
```diff
@@ -336,15 +459,254 @@ def search_incidents(client, args):
...
+                field=field, old_value=incident[field], new_value=new_value
+            )
+        )
 
+    dto = {"changes": changes}
+    demisto.debug(f"prepare_incident_update_dto_for_mirror {dto=}")
+    return dto
+
+
+# def to_timestamp(date_str_or_dt, date_format='%Y-%m-%dT%H:%M:%SZ'):
```
Comment (thefrieddan1) [inline Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py]: ?

File: Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py [hunk 8]
```diff
@@ -336,15 +459,254 @@ def search_incidents(client, args):
...
+#             parsed_time = time.strptime(date_str_or_dt, date_format)
+#             return int(time.mktime(parsed_time) * 1000)
+#
+#     # If the input is a datetime object, convert it to a timestamp
+#     if isinstance(date_str_or_dt, datetime):
+#         return int(time.mktime(date_str_or_dt.timetuple()) * 1000)
+#
+#     raise TypeError("Unsupported type for date_str_or_dt")
+
+from datetime import datetime, timezone
```
Comment (thefrieddan1) [inline Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py]: why import here?

File: Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py [hunk 9]
```diff
@@ -922,6 +1334,35 @@ def incident_attachments(client, incident_id):
...
+    try:
+        file_path_obj = demisto.getFilePath(entry_id)
+    except ValueError:
+        return CommandResults(
+            entry_type=EntryType.ERROR,
+            readable_output=f"Could not find a file with entry ID: {entry_id}",
+        )
+
+    file_path, file_name = file_path_obj.get("path"), file_path_obj.get("name")
+    if '.' in file_name:
```
Comment (thefrieddan1) [inline Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py:1316]: consider using os.path.splitext and then check if it has file extension or not.

File: Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py [hunk 10]
```diff
@@ -922,6 +1334,35 @@ def incident_attachments(client, incident_id):
...
+        tagged_file_name, extension = file_name.split('.')[0], file_name.split('.')[1]
+        file_name = f'{tagged_file_name}_{tag_to_ibm}.{extension}'
+    response = client.post_attachment(
+        uri=f"/incidents/{incident_id}/attachments",
+        filepath=file_path,
+        filename=file_name,
+    )
+    demisto.debug(f"upload_incident_attachment_command {response=}")
```
Comment (thefrieddan1) [inline Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py:1324]: no return statement.

File: Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py [hunk 11]
```diff
@@ -980,28 +1421,107 @@ def related_incidents(client, incident_id):
...
+                     f'with IDs: {[incident.get("id") for incident in resilient_incidents]}')
+        #  Update last_run_time to the latest incident creation time (maximum in milliseconds).
+        last_fetched_timestamp = last_incident_creation_time = max(
+            [_incident.get("create_date") for _incident in resilient_incidents]
+        )
+        demisto.debug(f'fetch_incidents {last_incident_creation_time=}')
+        for incident in resilient_incidents:
+
+            # Only fetching non-resolved incidents if `fetch_closed` is disabled.
+            if fetch_closed or (not incident.get('end_date') and incident.get('plan_status') == 'A'):  # 'A' stands for 'Active'
```
Comment (thefrieddan1) [inline Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py:1421]: consider using an Enum Active with value of A it adds readablity and reduces the comment.
something like:
class PlanStatus(Enum):
    ACTIVE = 'A'
    CLOSED = 'C'

File: Packs/IBMResilientSystems/Scripts/IbmUpdateNote/IbmUpdateNote.py (hunk 2 above)
Comment (barryyosi-panw) [reply Packs/IBMResilientSystems/Scripts/IbmUpdateNote/IbmUpdateNote.py:10]: yes, thanks

File: Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py (hunk 6 above)
Comment (barryyosi-panw) [reply Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py:179]: anyways, this constant was not used.

File: Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py (hunk 7 above)
Comment (barryyosi-panw) [reply Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py]: fixed

File: Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py (hunk 9 above)
Comment (barryyosi-panw) [reply Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py:1316]: adapted the code accordingly with os.path.splitext

File: Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py (hunk 10 above)
Comment (barryyosi-panw) [reply Packs/IBMResilientSystems/Integrations/IBMResilientSystems/IBMResilientSystems.py:1324]: there shouldn't necessarily be, it is not a command function. just executing the file upload.

**Review by thefrieddan1 — APPROVED** [review]

File: Packs/IBMResilientSystems/Scripts/IbmConvertCommentsToTable/IbmConvertCommentsToTable.yml [hunk 12]
```diff
@@ -0,0 +1,16 @@
+commonfields:
+  id: IbmConvertCommentsToTable
+  version: -1
+comment: This script is used to convert IBM QRadar SOAR notes to a markdown table.
```
Comment (ShirleyDenkberg) [inline Packs/IBMResilientSystems/Scripts/IbmConvertCommentsToTable/IbmConvertCommentsToTable.yml:4]: Should this be comments?
This script is used to convert IBM QRadar SOAR comments to a markdown table.

File: Packs/IBMResilientSystems/Scripts/IbmConvertCommentsToTable/README.md [hunk 13]
```diff
@@ -0,0 +1,21 @@
+This script is used to format IBM QRadar SOAR notes into a mark-down table.
```
Comment (ShirleyDenkberg) [inline Packs/IBMResilientSystems/Scripts/IbmConvertCommentsToTable/README.md:1]: Should this be comments? (and it should be markdown not mark-down)
This script is used to convert IBM QRadar SOAR comments to a markdown table.

File: Packs/IBMResilientSystems/Scripts/IbmConvertTasksToTable/IbmConvertTasksToTable.yml [hunk 14]
```diff
@@ -0,0 +1,16 @@
+commonfields:
+  id: IbmConvertTasksToTable
+  version: -1
+comment: This script is used to convert IBM QRadar SOAR notes to a markdown table.
```
Comment (ShirleyDenkberg) [inline Packs/IBMResilientSystems/Scripts/IbmConvertTasksToTable/IbmConvertTasksToTable.yml:4]: Should this be tasks?
This script is used to convert IBM QRadar SOAR tasks to a markdown table.

File: Packs/IBMResilientSystems/Scripts/IbmConvertTasksToTable/README.md [hunk 15]
```diff
@@ -0,0 +1,21 @@
+This script is used to format IBM QRadar SOAR tasks into a mark-down table.
```
Comment (ShirleyDenkberg) [inline Packs/IBMResilientSystems/Scripts/IbmConvertTasksToTable/README.md:1]:
```suggestion
This script is used to format IBM QRadar SOAR tasks into a markdown table.
```

File: Packs/IBMResilientSystems/Scripts/IbmUpdateNote/IbmUpdateNote.yml [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+  name: note_body
+  required: true
+
+- auto: PREDEFINED
+  description: The note tag. Use the note entry tag (defined in your instance configuration) to mirror the note to IBM QRadar SOAR.
+  name: tags
+  required: true
+  isArray: true
+
+comment: Use this script to add a note entry in XSOAR, which will then be mirrored as a note to an IBM QRadar SOAR incident. This script should be run within an incident.
```
Comment (ShirleyDenkberg) [inline Packs/IBMResilientSystems/Scripts/IbmUpdateNote/IbmUpdateNote.yml:16]:
```suggestion
comment: Use this script to add a note entry in Cortex XSOAR, which will then be mirrored as a note to an IBM QRadar SOAR incident. This script should be run within an incident.
```

File: Packs/IBMResilientSystems/Scripts/IbmUpdateTask/IbmUpdateTask.yml [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+  - Post-Incident
+  - Custom
+  - Complete
+- name: status
+  description: Changing the status field, completes or re-opens the task.
+  auto: PREDEFINED
+  predefined:
+  - Open
+  - Completed
+comment: Use this script to add a note entry in XSOAR, which will then be mirrored as a note to an IBM QRadar SOAR incident. This script should be run within an incident.
```
Comment (ShirleyDenkberg) [inline Packs/IBMResilientSystems/Scripts/IbmUpdateTask/IbmUpdateTask.yml:31]:
```suggestion
comment: Use this script to add a note entry in Cortex XSOAR, which will then be mirrored as a note to an IBM QRadar SOAR incident. This script should be run within an incident.
```

**Comment by ShirleyDenkberg** [comment]
@thefrieddan1 @altmannyarden @idovandijk @sapirshuker Doc review completed.

**Pull request merged**